File: TypeScript/143.重排链表 辅助数组.ts
// Definition for singly-linked list.
class ListNode {
    val: number
    next: ListNode | null
    constructor(val?: number, next?: ListNode | null) {
        this.val = (val === undefined ? 0 : val)
        this.next = (next === undefined ? null : next)
    }
}

function reorderList(head: ListNode | null): void {
    if (head === null) {
        return
    }
    const arr = new Array<ListNode>()
    while (head !== null) {
        arr.push(head)
        head = head.next
    }
    let [left, right] = [0, arr.length - 1]
    while (left < right) {
        arr[left].next = arr[right]
        ++left
        if (left == right) {
            break
        }
        arr[right].next = arr[left]
        --right
    }
    arr[left].next = null
};

File: Golang/143.重排链表 辅助数组.go
package leetcode

// Definition for singly-linked list.
type ListNode struct {
	Val  int
	Next *ListNode
}

func reorderList(head *ListNode) {
	if head == nil {
		return
	}
	arr := []*ListNode{}
	for head != nil {
		arr = append(arr, head)
		head = head.Next
	}
	left, right := 0, len(arr)-1
	for left < right {
		arr[left].Next = arr[right]
		left++
		if left == right {
			break
		}
		arr[right].Next = arr[left]
		right--
	}
	arr[left].Next = nil
}
